### PR TITLE
[SPARK-33497][SQL] Override maxRows in some LogicalPlan

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

This PR aims to override maxRows method in these follow `LogicalPlan`:
* `ReturnAnswer`
* `Join`
* `Range`
* `Sample`
* `RepartitionOperation`
* `Deduplicate`
* `LocalRelation`
* `Window`

### Why are the changes needed?

1. Logically, we know the max rows info with these `LogicalPlan`. 
2. Before this PR, we already have some max rows with `LogicalPlan`, so we can eliminate limit with more case if we expand more.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Add test.

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsSuite.scala [hunk 1]
```diff
@@ -48,8 +49,18 @@ class EliminateSortsSuite extends AnalysisTest {
...
+    Seq(
+      Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3),
+      Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3))
+  )
+  val testRelationB = LocalRelation.fromExternalRows(
+    Seq(Symbol("d").int),
+    Seq(
+      Row(1), Row(1), Row(1), Row(1), Row(1), Row(1),
+      Row(1), Row(1), Row(1), Row(1), Row(1), Row(1))
+  )
```
Comment (ulysses-you) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsSuite.scala:63]: This test uses rule `LimitPushDown`, so fill some data to avoid test fail.

**Comment by ulysses-you** [comment]
cc @maropu @cloud-fan

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala [hunk 2]
```diff
@@ -39,7 +41,18 @@ class CombiningLimitsSuite extends PlanTest {
         SimplifyConditionals) :: Nil
   }
 
-  val testRelation = LocalRelation('a.int, 'b.int, 'c.int)
+  val testRelation = LocalRelation.fromExternalRows(
+    Seq(Symbol("a").int, Symbol("b").int, Symbol("c").int),
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala:45]: nit: `'a.int, 'b.int, 'c.int`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala [hunk 3]
```diff
@@ -39,7 +41,18 @@ class CombiningLimitsSuite extends PlanTest {
...
+    Seq(
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala:46]: nit: 1.to(10).map(_ => Row(1, 2, 3))

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala [hunk 4]
```diff
@@ -117,4 +130,85 @@ class CombiningLimitsSuite extends PlanTest {
...
+    // test Range
+    checkPlan(
+      Range(1, 100, 1, None).select().limit(200),
+      Range(1, 100, 1, None).select()
+    )
+
+    // test Sample
+    val sampleQuery = testRelation.select().sample(upperBound = 0.2).limit(10).analyze
+    val sampleOptimized = Optimize.execute(sampleQuery)
+    assert(sampleOptimized.collect { case l @ Limit(_, _) => l }.isEmpty)
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala:154]: why can't we use `checkPlan` here?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala [hunk 5]
```diff
@@ -117,4 +130,85 @@ class CombiningLimitsSuite extends PlanTest {
...
+
+    // test Deduplicate
+    checkPlan(
+      testRelation.deduplicate(Symbol("a")).limit(10),
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala:158]: nit: `'a`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsSuite.scala [hunk 6]
```diff
@@ -48,8 +49,18 @@ class EliminateSortsSuite extends AnalysisTest {
       Batch("Limit PushDown", FixedPoint(10), LimitPushDown) :: Nil
   }
 
-  val testRelation = LocalRelation('a.int, 'b.int, 'c.int)
-  val testRelationB = LocalRelation('d.int)
+  val testRelation = LocalRelation.fromExternalRows(
+    Seq(Symbol("a").int, Symbol("b").int, Symbol("c").int),
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsSuite.scala:53]: ditto

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/LimitPushdownSuite.scala [hunk 7]
```diff
@@ -38,8 +39,12 @@ class LimitPushdownSuite extends PlanTest {
         BooleanSimplification) :: Nil
   }
 
-  private val testRelation = LocalRelation('a.int, 'b.int, 'c.int)
-  private val testRelation2 = LocalRelation('d.int, 'e.int, 'f.int)
+  private val testRelation = LocalRelation.fromExternalRows(
+    Seq(Symbol("a").int, Symbol("b").int, Symbol("c").int),
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/LimitPushdownSuite.scala:43]: ditto

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala (hunk 4 above)
Comment (ulysses-you) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala:154]: The `Sample` has a seed which is rand.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala (hunk 5 above)
Comment (ulysses-you) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala:156]: @cloud-fan Is these case needed because of [SPARK-29392](https://issues.apache.org/jira/browse/SPARK-29392) ?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala (hunk 3 above)
Comment (ulysses-you) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala]: Updated.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala [hunk 8]
```diff
@@ -452,6 +453,16 @@ package object dsl {
 
       def hint(name: String, parameters: Any*): LogicalPlan =
         UnresolvedHint(name, parameters, logicalPlan)
+
+      def sample(
+          lowerBound: Double = 0.0,
+          upperBound: Double,
+          withReplacement: Boolean = false,
+          seed: Long = Utils.random.nextLong): LogicalPlan = {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala:461]: Why do some params above have default values?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala (hunk 4 above)
Comment (maropu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala:152]: How about passing a `seed` param into `sample` and then using `checkPlan`?

Comment (ulysses-you) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala]: It's a way, updated.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 9]
```diff
@@ -541,6 +561,14 @@ case class Range(
     s"Range ($start, $end, step=$step, splits=$numSlices)"
   }
 
+  override def maxRows: Option[Long] = {
+    if (numElements.isValidLong) {
+      Some(numElements.toLong)
+    } else {
+      None
```
[the remark this answers is not in the document]
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:568]: Add the test.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 10]
```diff
@@ -958,6 +987,7 @@ case class Distinct(child: LogicalPlan) extends UnaryNode {
 abstract class RepartitionOperation extends UnaryNode {
   def shuffle: Boolean
   def numPartitions: Int
+  override def maxRows: Option[Long] = child.maxRows
```
[the remark this answers is not in the document]
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Added.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala (hunk 8 above)
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala:461]: Just follow the `Dataset.sample` and easy to use.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala [hunk 11]
```diff
@@ -117,4 +128,90 @@ class CombiningLimitsSuite extends PlanTest {
       testRelation.select().groupBy()(count(1)).orderBy(count(1).asc).analyze)
     comparePlans(optimized4, expected4)
   }
+
+  test("SPARK-33497: Override maxRows in some LogicalPlan") {
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala:132]: Could you split this test into pieces then add clearer test title for them?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala [hunk 12]
```diff
@@ -117,4 +128,90 @@ class CombiningLimitsSuite extends PlanTest {
...
+    checkPlan(
+      testRelation.window(
+        Seq(count(1).as("c")), Seq(Symbol("a")), Seq(Symbol("b").asc)).limit(20),
+      testRelation.window(
+        Seq(count(1).as("c")), Seq(Symbol("a")), Seq(Symbol("b").asc))
+    )
+  }
+}
+
+case class InfiniteRelation(output: Seq[Attribute]) extends LeafNode {
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala:211]: Infinite? How about `RelationWithoutMaxRows`?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamSuite.scala [hunk 13]
```diff
@@ -1130,7 +1130,7 @@ class StreamSuite extends StreamTest {
     verifyLocalLimit(inputDF.dropDuplicates().repartition(1).limit(1), expectStreamingLimit = false)
 
     // Should be LocalLimitExec in the first place, not from optimization of StreamingLocalLimitExec
-    val staticDF = spark.range(1).toDF("value").limit(1)
+    val staticDF = spark.range(2).toDF("value").limit(1)
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamSuite.scala:1133]: Why did you change the number?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala (hunk 11 above)
Comment (maropu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala:132]: Also, could you add tests to check if an expected max row value is propagated correctly?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala (hunk 5 above)
Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala:161]: ah I see. It's weird that the previous code still uses `'a`... Anyway let's use Symbol now.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamSuite.scala (hunk 13 above)
Comment (ulysses-you) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamSuite.scala:1133]: This case test the `LocalLimit`, and `range(1).limit(1)` will eliminate the limit. So here we change to range(2).

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala (hunk 12 above)
Comment (ulysses-you) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala]: Updated

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala (hunk 11 above)
Comment (ulysses-you) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala]: Yes, it will better.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala [hunk 14]
```diff
@@ -117,4 +128,111 @@ class CombiningLimitsSuite extends PlanTest {
...
+    checkPlan(
+      testRelation.window(
+        Seq(count(1).as("c")), Seq(Symbol("a")), Seq(Symbol("b").asc)).limit(20),
+      testRelation.window(
+        Seq(count(1).as("c")), Seq(Symbol("a")), Seq(Symbol("b").asc)),
+      10
+    )
+  }
+
+  private def checkPlan(
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala:225]: nit: `checkPlan` -> `checkPlanAndMaxRow`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala [hunk 15]
```diff
@@ -117,4 +128,111 @@ class CombiningLimitsSuite extends PlanTest {
...
+      10
+    )
+    checkPlan(
+      testRelation.distribute(Symbol("a"))(2).limit(10),
+      testRelation.distribute(Symbol("a"))(2),
+      10
+    )
+  }
+
+  test("SPARK-33497: Eliminate Limit if Join max rows not larger than Limit") {
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala:182]: It seems this test unit has similar tests, so could you make it more compact just like this?
```
Seq(Inner, FullOuter, LeftOuter, ...).foreach { joinType =>
  checkPlan(...)
}
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 16]
```diff
@@ -326,6 +327,25 @@ case class Join(
...
 
+  override def maxRows: Option[Long] = {
+    joinType match {
+      case Inner | Cross | FullOuter | LeftOuter | RightOuter
+        if left.maxRows.isDefined && right.maxRows.isDefined =>
+        val maxRows = BigInt(left.maxRows.get) * BigInt(right.maxRows.get)
+        if (maxRows.isValidLong) {
+          Some(maxRows.toLong)
+        } else {
+          None
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:338]: Could you add tests for this case in `test("SPARK-33497: Eliminate Limit if Join max rows not larger than Limit")`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala (hunk 14 above)
Comment (ulysses-you) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala]: Updated

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala (hunk 15 above)
Comment (ulysses-you) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala:182]: yeah it looks better, thanks

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 16 above)
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:338]: Added by
```
  // maxRow is not valid long
  checkPlanAndMaxRow(
    testRelation.join(testRelation4, joinType).limit(100),
    testRelation.join(testRelation4, joinType).limit(100),
    100
  )
```

**Review by maropu — APPROVED** [review]

**Comment by ulysses-you** [comment]
cc @cloud-fan

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsSuite.scala [hunk 17]
```diff
@@ -48,8 +49,18 @@ class EliminateSortsSuite extends AnalysisTest {
...
+    Seq(
+      Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3),
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsSuite.scala:55]: ditto, `1.to(10).map(_ => Row(1, 2, 3)`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsSuite.scala [hunk 18]
```diff
@@ -48,8 +49,18 @@ class EliminateSortsSuite extends AnalysisTest {
...
+      Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3))
+  )
+  val testRelationB = LocalRelation.fromExternalRows(
+    Seq(Symbol("d").int),
+    Seq(
+      Row(1), Row(1), Row(1), Row(1), Row(1), Row(1),
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsSuite.scala:61]: ditto

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/LimitPushdownSuite.scala [hunk 19]
```diff
@@ -38,8 +39,12 @@ class LimitPushdownSuite extends PlanTest {
...
+    Seq(Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3)))
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/LimitPushdownSuite.scala:44]: ditto

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/LimitPushdownSuite.scala [hunk 20]
```diff
@@ -38,8 +39,12 @@ class LimitPushdownSuite extends PlanTest {
...
+  private val testRelation2 = LocalRelation.fromExternalRows(
+    Seq(Symbol("d").int, Symbol("e").int, Symbol("f").int),
+    Seq(Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3), Row(1, 2, 3)))
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/LimitPushdownSuite.scala:47]: ditto

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsSuite.scala (hunk 17 above)
Comment (ulysses-you) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsSuite.scala]: Updated these.

**Comment by ulysses-you** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala [hunk 21]
```diff
@@ -452,6 +453,16 @@ package object dsl {
 
       def hint(name: String, parameters: Any*): LogicalPlan =
         UnresolvedHint(name, parameters, logicalPlan)
+
+      def sample(
+          lowerBound: Double = 0.0,
+          upperBound: Double,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala:459]: It's pretty weird that a parameter without default value comes after a parameter with default value. If we want to follow Dataset, let's use overloads. Otherwise, let's make this `sample` method simple and good enough for the added tests.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala [hunk 22]
```diff
@@ -117,4 +128,118 @@ class CombiningLimitsSuite extends PlanTest {
...
+      testRelation.select().sample(upperBound = 0.2, seed = 1).limit(10),
+      testRelation.select().sample(upperBound = 0.2, seed = 1),
+      10
+    )
+  }
+
+  test("SPARK-33497: Eliminate Limit if Deduplicate max rows not larger than Limit") {
+    checkPlanAndMaxRow(
+      testRelation.deduplicate(Symbol("a")).limit(10),
+      testRelation.deduplicate(Symbol("a")),
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala:164]: I just found out a better way: `"a".attr`. Can you make a try and see if it can replace `Symbol("a")`?

Comment (ulysses-you) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CombiningLimitsSuite.scala]: Add a new implicit `DslAttr` in dsl, seems better.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala (hunk 21 above)
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala:461]: OK, let's keep `sample` simple.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by ulysses-you** [comment]
thanks @cloud-fan @maropu
